alternator/executor.cc

```diff
@@ -3145,19 +3145,58 @@ future<executor::request_return_type> executor::batch_get_item(client_state& cli
         rjson::add(response, "UnprocessedKeys", rjson::empty_object());
-        for (auto& t : responses) {
-            if (!response["Responses"].HasMember(std::get<0>(t).c_str())) {
-                rjson::add_with_string_name(response["Responses"], std::get<0>(t), rjson::empty_array());
+        // In case of full failure (no reads succeeded), an arbitrary error
+        // from one of the operations will be returned.
+        bool some_succeeded = false;
+        std::exception_ptr eptr;
+        // These iterators are used to match keys from the requests with their corresponding responses.
+        // If any of the responses failed, the key iterator will be used to insert
+        // an entry into the UnprocessedKeys object, which will ultimately be returned
+        // to the user in case of partial success of BatchGetItem operation.
+        rjson::value::MemberIterator item_it = request_items.MemberBegin();
+        rjson::value::ValueIterator key_it = item_it->value["Keys"].Begin();
+
+        for (auto& fut : responses) {
+            if (fut.failed()) {
+                eptr = fut.get_exception();
+                if (!response["UnprocessedKeys"].HasMember(item_it->name)) {
+                    rjson::add_with_string_name(response["UnprocessedKeys"], rjson::to_string_view(item_it->name), rjson::empty_object());
+                    rjson::value& unprocessed_item = response["UnprocessedKeys"][item_it->name];
+                    for (auto it = item_it->value.MemberBegin(); it != item_it->value.MemberEnd(); ++it) {
+                        if (it->name != "Keys") {
+                            rjson::add_with_string_name(unprocessed_item, rjson::to_string_view(it->name), rjson::copy(it->value));
+                        }
+                    }
+                    rjson::add_with_string_name(unprocessed_item, "Keys", rjson::empty_array());
+                }
+                rjson::push_back(response["UnprocessedKeys"][item_it->name]["Keys"], std::move(*key_it));
+            } else {
+                auto t = fut.get();
+                some_succeeded = true;
+                if (!response["Responses"].HasMember(std::get<0>(t).c_str())) {
+                    rjson::add_with_string_name(response["Responses"], std::get<0>(t), rjson::empty_array());
+                }
+                if (std::get<1>(t)) {
+                    rjson::push_back(response["Responses"][std::get<0>(t)], std::move(*std::get<1>(t)));
+                }
             }
-            if (std::get<1>(t)) {
-                rjson::push_back(response["Responses"][std::get<0>(t)], std::move(*std::get<1>(t)));
+            key_it++;
+            if (key_it == item_it->value["Keys"].End()) {
+                item_it++;
+                if (item_it != request_items.MemberEnd()) {
+                    key_it = item_it->value["Keys"].Begin();
+                }
             }
         }
+        elogger.debug("Unprocessed keys: {}", response["UnprocessedKeys"]);
```

Comment:
do you want to keep this in the code?

Reply:
Not sure. It's definitely useful for testing the code manually, but it might be a little too spammy for debug. On the other hand, `trace` log level is where potentially spammy messages belong, so maybe I'll just degrade it to `trace`.